presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java

```diff
@@ -1121,10 +1108,159 @@ protected Scope visitTable(Table table, Optional<Scope> scope)
+        }
+
+        private Scope processMaterializedView(
+                Table materializedView,
+                QualifiedObjectName materializedViewObjectName,
+                Optional<Scope> scope,
+                QualifiedObjectName materializedViewQualifiedObjectName,
+                ConnectorMaterializedViewDefinition materializedViewDefinition,
+                Statement statement)
+        {
+            MaterializedViewStatus materializedViewStatus = metadata.getMaterializedViewStatus(session, materializedViewObjectName);
+
+            String newSql = getMaterializedViewSQL((Query) statement, materializedViewDefinition, materializedViewStatus);
+            System.out.println("New generated SQL is: \n" + newSql);
+
+            Query query = (Query) sqlParser.createStatement(newSql, createParsingOptions(session, warningCollector));
+            analysis.registerNamedQuery(materializedView, query);
+
+            analysis.registerMaterializedViewForAnalysis(materializedView);
+            RelationType relationType = analyzeView(query, materializedViewQualifiedObjectName, Optional.of(materializedViewObjectName.getCatalogName()), Optional.of(materializedViewDefinition.getSchema()), materializedViewDefinition.getOwner(), materializedView);
+            analysis.unregisterMaterializedViewForAnalysis(materializedView);
+            return createAndAssignScope(materializedView, scope, relationType);
+        }
+
+        private String getMaterializedViewSQL(Query statement, ConnectorMaterializedViewDefinition connectorMaterializedViewDefinition, MaterializedViewStatus materializedViewStatus)
+        {
+            String materializedViewCreateSql = connectorMaterializedViewDefinition.getOriginalSql();
+            if (materializedViewStatus.isNotMaterialized()) {
+                return materializedViewCreateSql;
+            }
+
+            QuerySpecification userQuerySpecification = (QuerySpecification) statement.getQueryBody();
+            if (materializedViewStatus.getMaterializedViewState() == FULLY_MATERIALIZED) {
+                Query query = new Query(Optional.empty(), userQuerySpecification, Optional.empty(), Optional.empty());
+                return SqlFormatterUtil.getFormattedSql(query, sqlParser, Optional.empty());
+            }
+
+            checkState(materializedViewStatus.isPartiallyMaterialized(), "Materialized view should be partially materialized.");
+
+            Statement createSqlStatement = sqlParser.createStatement(materializedViewCreateSql, createParsingOptions(session, warningCollector));
+            QuerySpecification createSqlQuerySpecification = (QuerySpecification) ((Query) createSqlStatement).getQueryBody();
+
+            Map<SchemaTableName, Expression> partitionPredicates = generatePartitionPredicate(materializedViewStatus.getPartitionsFromBaseTables());
+
+            Node predicateStitchedNode = new PredicateStitcher(session, partitionPredicates).process(createSqlQuerySpecification);
```

Comment:
I'm curious about the following statement:
```sql
CREATE MATERIALIZED VIEW m
WITH (partitioned_by = ARRAY['k1', 'k2'])
AS
SELECT A.k1, B.k2
FROM A LEFT JOIN B
ON A.k1 = B.K2
```

Suppose `A.k1` has values `{'p1', 'p2'}` and `B.k2` has values `{'p1'}`. Then `m` will have partitions `k1=p1/k2=p1` and `k1=p2/k2=null`. This would be an issue right? Though we guarantee strong partition mapping, we can still produce partition values that are not present in the original base tables (which in this case is `null`).

If that is the case, I would be curious if supporting join like this is valid or not.

Reply:
Discussed offline, we will not support any other join except inner for now. We will add query plan validation in the MV validation.